DeploymentAgentReconfigurationScripts/TFS2018 Update2/HardwareMoveRemoveOldAgent.ps1

```diff
@@ -0,0 +1,178 @@
+$deploymentMachineId = $dm.value[0].id;
+
+if ($dm.value[0].agent.status -ne 'offline'){
+    Write-Verbose -Verbose "Agent is already configured to the TFS $targetTFSUrl"
+    return 0;
+}
+
+
+# Create and go to new agent folder
+If(-NOT (Test-Path $env:SystemDrive\'vstsagent'))
+{
+    mkdir $env:SystemDrive\'vstsagent'
+}
+
+$newAgentFolfer = "";
+
+cd $env:SystemDrive\'vstsagent';
+for ($i=1; $i -lt 100; $i++){
+    $destFolder="A"+$i.ToString();
+    if (-NOT (Test-Path ($destFolder))){
+        $newAgentFolfer = $destFolder;
+        break;
+    }
+};
+$newAgentPath = $env:SystemDrive + '\vstsagent\' + $newAgentFolfer;
+
+if ($action -ne "apply"){
+    Write-Verbose -Verbose "If action is set to apply, this script will delete the existing unusable, offline agent $agentName from the deployment group $deploymentGroupName in the TFS $targetTFSUrl and re-configure a new agent in $newAgentPath path with same name and properties including tags $tags. All the old deployment histroty will be deleted.";
+    return 0;
+}
+
+Write-Verbose -Verbose "Start execution : It will delete the existing unusable, offline agent $agentName from the deployment group $deploymentGroupName in the TFS $targetTFSUrl and re-configure a new agent in $newAgentPath path with same name and properties including tags $tags. All the old deployment histroty will be deleted.";
+
+mkdir $newAgentFolfer;
+cd $newAgentFolfer; 
+
+# download the agent bits.
+$agentZip= $agentZip="$PWD\agent.zip";
+$DefaultProxy=[System.Net.WebRequest]::DefaultWebProxy;$securityProtocol=@();
+$securityProtocol+=[Net.ServicePointManager]::SecurityProtocol;
+$securityProtocol+=[Net.SecurityProtocolType]::Tls12;[Net.ServicePointManager]::SecurityProtocol=$securityProtocol;
+$WebClient=New-Object Net.WebClient;
+    
+if($DefaultProxy -and (-not $DefaultProxy.IsBypassed($agentDownloadUrl)))
+{
+    $WebClient.Proxy= New-Object Net.WebProxy($DefaultProxy.GetProxy($agentDownloadUrl).OriginalString, $True);
+};
+
+$WebClient.DownloadFile($agentDownloadUrl, $agentZip);
+
+Add-Type -AssemblyName System.IO.Compression.FileSystem;[System.IO.Compression.ZipFile]::ExtractToDirectory( $agentZip, "$PWD"); 
+
+# Delete existing agent reference from the target TFS	
+$deleteDeploymentMachineUrl = $targetTFSUrl + '/' + $collectionName + '/' + $projectName + '/_apis/distributedtask/deploymentgroups/' + $deploymentGroupId + '/targets/' + $deploymentMachineId + '?api-version=5.0-preview.1'
+Invoke-RestMethod -Uri $deleteDeploymentMachineUrl -Method Delete -Headers @{Authorization = "Basic $encodedPat"} -ContentType "application/json"
+
+
+# Re-configure the agent to the target TFS
+if ($tags -ne ""){
+    .\config.cmd --deploymentgroup --url $targetTFSUrl --collectionname $collectionName --projectname $projectName --deploymentgroupname $deploymentGroupName --agent $agentName --auth Integrated --runasservice --work '_work' --unattended --adddeploymentgrouptags --deploymentgrouptags $tags
```

Comment:
do we want auth mode hard coded to "Integrated " always or this is a sample for user he can modify the script as per his scenario

Reply:
You are handling proxy in download but not during config.

Reply:
Handling proxy during configuration is complex scenario. May be user need to modify accordingly.  lets keep the sample script simple.